source/fundamentals/data-formats/codecs.txt

```diff
@@ -11,38 +11,360 @@ Codecs
+implementations:
+
+- :java-docs:`IntegerCodec </apidocs/bson/org/bson/codecs/IntegerCodec.html>`,
+  a ``Codec`` that converts Integers and is part of the BSON package.
+- :ref:`PowerStatusCodec <codecs-powerstatus-codec>`, our sample ``Codec``
+  that converts certain Java strings to BSON booleans.
+
+You can retrieve the ``Codec`` instances from the ``CodecRegistry`` instance
+from the prior example using the following code:
+
+.. code-block:: java
+   :copyable: true
+
+   Codec<String> powerStatusCodec = codecRegistry.get(String.class);
+   Codec<Integer> integerCodec = codecRegistry.get(Integer.class);
+
+If you attempt to retrieve a ``Codec`` instance for a class that is not
+registered, the ``get()`` method throws a ``CodecConfigurationException``
+exception.
+
+For more information about the classes and interfaces in this section, see the
+following API documentation:
+
+- :java-docs:`CodecRegistries </apidocs/bson/org/bson/codecs/configuration/CodecRegistries.html>`
+- :java-docs:`IntegerCodec </apidocs/bson/org/bson/codecs/IntegerCodec.html>`
+
+.. _codecs-codecprovider:
 
 CodecProvider
 -------------
 
-- Description
-- Example of get method implementation
+A ``CodecProvider`` is an interface that contains abstract methods that create
+``Codec`` instances and assign them to a ``CodecRegistry`` instance. Similar
+to the ``CodecRegistry``, the BSON library uses the ``Codec`` instances
+retrieved by the ``get()`` method to convert between Java and BSON data types.
+
+However, in cases in which you add a class that contains fields that require
+corresponding ``Codec`` objects, you need to ensure that you instantiate the
+``Codec`` objects for the class' fields before you instantiate the
+``Codec`` for the class. You can use the ``CodecRegistry`` parameter in
+the ``get()`` method to pass any of the ``Codec`` instances that the
+``Codec`` relies on.
+
+The following code example shows how you can implement ``CodecProvider`` to
+pass the ``MonolightCodec`` any ``Codec`` instances it needs in a
+``CodecRegistry`` instance such as the ``PowerStatusCodec`` from our prior
+example:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/MonolightCodecProvider.java
+   :start-after: start class
+   :end-before: end class
+   :language: java
+   :dedent:
+
+To see a runnable example that demonstrates read and write operations using
+these ``Codec`` classes, see the :ref:`Custom Codec Example <codecs-custom-example>`
+section of this guide.
+
+When working with POJOs, consider using the  ``PojoCodecProvider`` to
+minimize duplicate code to convert commonly-used data types and customize
+the behavior. See our
+:doc:`POJO Customization guide </fundamentals/data-formats/pojo-customization>`
+for more information.
+
+For more information about the classes and interfaces in this section, see the
+:java-docs:`CodecProvider API documentation <apidocs/bson/org/bson/codecs/configuration/CodecProvider.html>`.
 
 BsonTypeClassMap
```

Comment:
This section doesn't need to be nested in CodecProvider.

`BsonTypeClassMap` is really an implementation detail for Codec's that contain container classes like `Document`, `Map` & `Iterables` and its only used for decoding of types from Bson into these container types.  Perhaps add a clarification at the start?

Reply:
Good to know, updating the section.